### PR TITLE
Remove BooleanHamiltonian object

### PR DESCRIPTION
**Pull request by tonybruguier**
Now that v0.14 is released, delete deprecate object

### CODE REVIEW

**Comment by MichaelBroughton** [comment]
v0.14 isn't out just yet. Do we want to remove this before 0.14 or after ?

**Comment by tonybruguier** [comment]
I was planning to do it after 0.14 indeed. I was planning to flip the PR out of draft once 0.14 is released (I just had some time, so I decided to prepare it but not send it yet).

**Comment by tonybruguier** [comment]
@MichaelBroughton I think this is ready for review, but this is not urgent.

File: cirq-core/cirq/protocols/json_test_data/BooleanHamiltonian.json [hunk 1]
```diff
@@ -1,15 +0,0 @@
-[
-  {
-    "cirq_type": "BooleanHamiltonian",
```
Comment (tanujkhattar) [inline cirq-core/cirq/protocols/json_test_data/BooleanHamiltonian.json:3]: @MichaelBroughton IIUC, We do need to preserve the de-serialization of objects that were previously serializable even though the object itself is being removed? So deserializing `BooleanHamiltonian` should now result in instantiating a new `BooleanHamiltonianGate` object?

**Review by tanujkhattar — COMMENTED** [review]
LGTM % comment on whether we need to preserve de-serialization for deleted objects.

File: cirq-core/cirq/protocols/json_test_data/BooleanHamiltonian.json (hunk 1 above)
Comment (tonybruguier) [reply cirq-core/cirq/protocols/json_test_data/BooleanHamiltonian.json:3]: @tanujkhattar 
If Michael says to do so, how would I recover the qubits though? I think, if I am reading the code correctly, that the `BooleanHamiltonian` stores qubits, but the `BooleanHamiltonianGate` does not.

Comment (tanujkhattar) [reply cirq-core/cirq/protocols/json_test_data/BooleanHamiltonian.json:3]: The deserialization would return `BooleanHamiltonianGate(...).on(*qubits)`

Comment (tonybruguier) [reply cirq-core/cirq/protocols/json_test_data/BooleanHamiltonian.json]: Done.

**Comment by tonybruguier** [comment]
Thanks for having given me pointers during the meeting. I added de-serialization for deprecated objects. PTAL.

**Comment by tonybruguier** [comment]
Friendly ping. I think this is according to what was discussed at Cirq Cynq, but happy to revise further, of course. It's not super urgent, but maybe it's nice to clean up the code eventually?

**Review by tanujkhattar — APPROVED** [review]
LGTM

**Pull request merged**